### ISSUE TITLE
Use PulseAudio for BSD

### ISSUE THREAD

**Issue by hajimehoshi** [hajimehoshi/oto#154]
ALSA is not listed in OpenBSD package list

https://ftp.openbsd.org/pub/OpenBSD/7.0/packages/amd64/

**Comment by divVerent** [comment]
Yeah. FreeBSD has ALSA in the ports collection, and technically it wouldn't be much work to at least ship a pulseaudio-only-supporting ALSA userspace for OpenBSD or NetBSD... but it's not in their ports and pkgsrc.

I sure would like to see support for all the BSDs, though. Whether that means a pulseaudio or BSD audio / OSS-ish backend I do not really care.

**Comment by divVerent** [comment]
Actually, I may be wrong - apparently NetBSD has some ALSA stuff in its pkgsrc now. Gotta try that out. That would still leave OpenBSD though.

**Comment by midnadimple** [comment]
pulse is prob best option, since most people, regardless of what os they're using, will have it installed already. how difficult would it be to transition?

**Comment by hajimehoshi** [comment]
The implementation is not difficult. Actually there was an implementation https://github.com/hajimehoshi/ebiten/commit/13e84f022477ea6e4055b01438cf8f9390ffcb84

The problem is that I don't have an environment to test this.

**Comment by midnadimple** [comment]
huh, it just so happens that I do. ill test it out

**Comment by hajimehoshi** [comment]
Thanks, I'll queue this task to my task list... Stay tuned

**Comment by divVerent** [comment]
As latency can matter a lot for some games, maybe at least for some
transition period offer both ALSA and pulseaudio (if both compiled in),
similar to how we now have DirectX and OpenGL on Ebiten?

**Comment by hajimehoshi** [comment]
This sounds a good idea, but the affected environment is only FreeBSD, and this is originally not well tested to be honest. So, I think it is ok not to have a transition period.

**Comment by divVerent** [comment]
Sorry, I mean - if you implement pulseaudio support, then that would not
only cover all the BSDs, but Linux too.

So if that is the approach taken, why not in the long term move to that for
Linux too? Would also make snap and flatpak integration easier.

However after all that tends to happen on Linux audio, I would want to see
a transition mechanism for that.

**Comment by Kimapr** [comment]
I use exclusively PulseAudio on my Linux device, and the use of ALSA here makes AAAAXY play sound on the wrong device (I want it to play over HDMI but it plays on laptop speakers instead) with no obvious way to fix this.

**Comment by hajimehoshi** [comment]
I'm not sure the convention in Linux. ALSA is a kind of lower-level thing than PulseAudio, but probably am I misunderstanding?

**Comment by karlpip** [comment]
I would like to ask if it's possible to leave the ALSA implementation in the code in case you want to implement Pulseaudio support. So one can chose between them somehow, maybe with an ENV var?

Currently i am using Oto on a machine which does not need Pulseaudio because there is only one sound emitting program at a time. In my case installing Pulseaudio would only be unnecessary bloat and energy waste as it starts user services.

**Comment by hajimehoshi** [comment]
Enabling to choose PluseAudio sounds an interesting idea. Let me think...

**Comment by divVerent** [comment]
> I'm not sure the convention in Linux. ALSA is a kind of lower-level thing than PulseAudio, but probably am I misunderstanding?

It's complicated. It technically is, but PulseAudio can use ALSA as a backend and ALSA can use PulseAudio as a backend.

**Comment by reezer** [comment]
I hope this is helpful and might also be interesting to people coming here searching. There is a very cross-platform C library that does support pretty much all of Go's supported platforms called [miniaudio](https://miniaud.io/). There is a Go library interfacing with it, called [malgo](https://github.com/gen2brain/malgo). I am writing it because it's a great overview of what could be done and which platforms support which backends. Unless many other options it also doesn't require much to set it up and get it building on all the platforms.

Maybe something to steal ideas or even code (miniaudio is public domain) from.

**Comment by thiagokokada** [comment]
One thing that I found interesting is that there are some non-CGO implementations of PulseAudio, like this one: https://github.com/jfreymuth/pulse.

I am very interested in having a non-CGO option for Linux, because cross-compiling ALSA statically is painful in my experience. Also getting support for BSD would be nice.

But even if this does depends in CGO I would prefer to have it instead of ALSA, since in my experience ALSA in slightly strange ways when running user PulseAudio/PipeWire (not deal breaker in any means).

> So one can chose between them somehow, maybe with an ENV var?

I would prefer this to be done via flags, because as a developer of the app I would prefer to explicitly disable ALSA since I don't have enough time to test my app in both.

However if there is a way to change this at runtime it would also be nice, but I would prefer that I am in control (like if I want the user to control this via ENV variables, I implement the ENV var and pass to the initialization of Oto, for example).
